File: pi-extension/src/session/session_gate.test.ts
import { describe, expect, test } from "vitest";
import type { ClientMessage } from "../protocol/types.js";
import { validateClientSession } from "./session_gate.js";

describe("validateClientSession", () => {
  test("accepts non-session-scoped messages", () => {
    const msg: ClientMessage = { type: "ping", id: "ping-1" };

    expect(validateClientSession(msg, "session-a")).toEqual({ ok: true });
  });

  test("accepts matching session-scoped commands", () => {
    const msg: ClientMessage = {
      type: "session_sync",
      id: "sync-1",
      session_id: "session-a",
    };

    expect(validateClientSession(msg, "session-a")).toEqual({ ok: true });
  });

  test("rejects missing or stale session ids", () => {
    const stale: ClientMessage = {
      type: "cancel",
      id: "cancel-1",
      session_id: "session-b",
      target_id: "turn-1",
    };
    const missing = {
      type: "session_compact",
      id: "compact-1",
    } as ClientMessage;

    expect(validateClientSession(stale, "session-a")).toMatchObject({
      ok: false,
      code: "session_mismatch",
      currentSessionId: "session-a",
      receivedSessionId: "session-b",
    });
    expect(validateClientSession(missing, "session-a")).toMatchObject({
      ok: false,
      code: "session_mismatch",
      currentSessionId: "session-a",
      receivedSessionId: null,
    });
  });
});

File: app/test/data/sync/session_gate_test.dart
import 'package:app/data/sync/session_gate.dart';
import 'package:app/domain/entities/remote_session_ref.dart';
import 'package:app/protocol/protocol.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  const gate = SessionGate();
  const active = RemoteSessionRef(
    peerEpk: 'peer-1',
    roomId: 'main',
    sessionId: 'session-1',
  );

  test('accepts non-session control messages without an active session', () {
    final result = gate.accepts(Pong(inReplyTo: 'ping-1'), null);
    expect(result.accepted, isTrue);
  });

  test('rejects session-scoped messages when active session is unknown', () {
    final result = gate.accepts(
      AgentChunk(sessionId: 'session-1', inReplyTo: 'u1', delta: 'hello'),
      null,
    );
    expect(result.accepted, isFalse);
    expect(result.reason, 'active_session_unknown');
  });

  test('rejects missing session_id on session-scoped messages', () {
    final result = gate.accepts(
      AgentChunk(inReplyTo: 'u1', delta: 'hello'),
      active,
    );
    expect(result.accepted, isFalse);
    expect(result.reason, 'missing_session_id');
  });

  test('rejects mismatched session_id', () {
    final result = gate.accepts(
      SessionHistory(
        sessionId: 'session-2',
        inReplyTo: 'sync-1',
        sessionStartedAt: 1,
        events: const [],
        eos: true,
      ),
      active,
    );
    expect(result.accepted, isFalse);
    expect(result.reason, 'session_mismatch');
    expect(result.messageSessionId, 'session-2');
    expect(result.expectedSessionId, 'session-1');
  });

  test('accepts matching session_id', () {
    final result = gate.accepts(
      QueuedMessageState(sessionId: 'session-1', id: 'q1', text: 'next'),
      active,
    );
    expect(result.accepted, isTrue);
  });
}
